### ISSUE TITLE
Hard to read Tokio docs and code: Handle

### ISSUE THREAD

**Issue by fpgaminer** [tokio-rs/tokio#480]
I'm currently reading through the Tokio code to try and understand how it works under the hood.

I'm opening this issue to note a bit of confusion I ran into while doing so.  Hopefully by reporting what troubles I run into on my first read it can help to make the code more approachable for future travelers.  If this is not helpful, please let me know.  Otherwise I'm happy to open issues for other code comprehension related issues I run into.

Multiple types share the exact same name: `Handle`.

Here are various examples through the code where `Handle` is referenced:

* https://github.com/tokio-rs/tokio/blob/tokio-0.1.7/src/executor/current_thread/mod.rs#L421
* https://github.com/tokio-rs/tokio/blob/tokio-0.1.7/src/runtime/current_thread/runtime.rs#L75
* https://github.com/tokio-rs/tokio/blob/tokio-0.1.7/tokio-timer/src/timer/handle.rs#L36

The most egregious example for me first occurred when I was just trying to use Tokio and read the docs: https://docs.rs/tokio/0.1.7/tokio/net/struct.TcpListener.html#method.from_std

In my case I was attempting to use net2::TcpBuilder to build a custom listener and then use from_std to get it into Tokio.  So I looked up the docs for from_std ... and had no idea what a `Handle` was.  So I searched the docs for `handle` and got multiple hits.  Little did I know there are _many_ different types with that same name.  It was, and is, very confusing.

Basically, someone reading the code has no idea which of the many types `Handle` refers to at any given time.  The only way to know is to search the current file for `Handle` and find either a `use` or a local type declaration.

You can see that this becomes a problem even within the Tokio code itself: https://github.com/tokio-rs/tokio/blob/tokio-0.1.7/src/runtime/current_thread/runtime.rs#L2

That module needs to `use` a `Handle`, but the name conflicts with a local `Handle` type, so it has to rename it.

I think all of these types should be uniquely named.  `TimerHandle`, `RuntimeHandle`, etc.

Note that there are also other of these same types of issues.  For example `runtime::Runtime` and `runtime::current_thread::Runtime`.

**Comment by tobz** [comment]
@carllerche Do you think you'd be open to a PR for this?  Uniquely name all `Handle` types based on what they're for.  I think this is a definitely an API paper cut and I'd be willing to put in the work on it.

**Comment by tobz** [comment]
@fpgaminer Carl can speak more to it, but I believe the `runtime::Runtime` vs `runtime::current_thread::Runtime` thing is sort of a trade-off of making one the default vs simply exposing the other.

It could probably still manage to be made clearer, though!

**Comment by carllerche** [comment]
Generally, the strategy w/ names like `Handle` would be to include the module as part of the name. So, `timer::Handle` vs. just `Handle`.

That said, I agree it can be a bit confusing. However, doing a full rename would be a breaking change unless care is taken... I also don't know if it would be better to rename the struct or always use it w/ the mod.

Perhaps an issue can be open on https://github.com/rust-lang-nursery/api-guidelines for discussion on style.

**Comment by stjepang** [comment]
Relevant API guidelines issue: https://github.com/rust-lang-nursery/api-guidelines/issues/66

**Comment by carllerche** [comment]
Given the lack of conclusion, I think we should hold off on any changes until a consensus emerges.

For now, I would recommend as a rule of thumb to include the module when the struct name is ambiguous. So: `timer::Handle`.
